### DISCUSSION TITLE
Code generators

### DISCUSSION THREAD

**Discussion by pvdbosch** [belgif/rest-guide#147] [General]
In this discussion, you can share your ideas or experiences about code generators (like openapi-generator or swagger-codegen) that you use it to generate server or client code from OpenAPI files that follow the REST guide.

**Comment by pvdbosch** [comment]
This thread is meant as a continuation of the discussions on OpenAPI code generators in issue #25 .

**Comment by pvdbosch** [comment]
About the confusion around typeMappings/schemaMappings/importMappings mentioned in #25 :

* use schemaMappings if you want to map to your own model class instead of generating one (we've used it and seems to work)
* importMappings still generates a class. It seems to have been designed [for something else](https://github.com/OpenAPITools/openapi-generator/issues/11506#issuecomment-1051781681), though not clear to me for which use case. It's listed in [the documentation](https://openapi-generator.tech/docs/customization/#bringing-your-own-models) under "bring your own model" but that seems to be incorrect.
* typeMappings can be used to modify the name of the generated class, to something else then the one in the OpenAPI file. There are [similar configurations](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator-maven-plugin/README.md) to modify parameter names, property names, inline types, etc.

**Comment by pvdbosch** [comment]
Generated classes for OpenAPI enums can be difficult to map to domain classes in Java because their enum values aren't generated in UPPER_SNAKE_CASE, but UPPERCASE.

There's a related open issue: https://github.com/OpenAPITools/openapi-generator/issues/4837 ; with workaround using "x-enum-varnames".
Anyone also experiencing this difficulty; are there other solutions?

**Reply by jpraet** [reply]
@pvdbosch this issue has been addressed in the latest openapi-generator 7.7.0.

**Comment by jpraet** [comment]
The upgrade from openapi-generator 7.1.0 to 7.2.0 is causing some problems for us (using the jax-rs generator).

It places `@Valid` bean validations on generic type parameters. e.g. `private @Valid List<@Valid Partner> items;`.
This annotation location is only supported in Bean Validation 2.0 so is causing issues for older Java EE 7 applications.

Furthermore, in more recent Java EE 8 applications on JBoss it also seems to be causing issues:

`WELD-001503: Bean class which has interceptors cannot be declared final: class be.fgov.kszbcss.tad.admin.rest.model.PartnersRest$PartnersRestBuilderImpl`

**Reply by jpraet** [reply]
See https://github.com/OpenAPITools/openapi-generator/issues/17874 and https://github.com/OpenAPITools/openapi-generator/issues/17875.

**Reply by jpraet** [reply]
https://github.com/OpenAPITools/openapi-generator/issues/17875 has been addressed in latest openapi-generator 7.7.0.

Some migration notes:
* Since 7.3.0 operations and models that are marked as `x-internal: true` are skipped during code generation.
To disable this, you need to configure `<openapiNormalizer>REMOVE_X_INTERNAL=true</openapiNormalizer>`.
Yes, you need to set REMOVE_X_INTERNAL to true to not remove them. Very counter-intuitive.
* Since 7.7.0 enum values are generated in UPPER_SNAKE_CASE

**Reply by jpraet** [reply]
Nope, at CBSS we had to rollback to 7.1.0 again because of https://github.com/OpenAPITools/openapi-generator/issues/19188.

**Reply by jpraet** [reply]
Any other belgif members impacted by https://github.com/OpenAPITools/openapi-generator/issues/19188?

**Reply by pvdbosch** [reply]
We haven't had any complaints yet, but we're using spring boot more often and we're maintaining variants of the code generation templates (mustache files).

**Reply by jpraet** [reply]
https://github.com/OpenAPITools/openapi-generator/issues/19188 got addressed in 7.11.0, so at CBSS we are now migrating from 7.1.0 to 7.11.0.

Some migration notes:

* Since 7.3.0 operations and models that are marked as x-internal: true are skipped during code generation.
  To disable this, you need to configure <openapiNormalizer>REMOVE_X_INTERNAL=true</openapiNormalizer>.
  Yes, you need to set REMOVE_X_INTERNAL to true to not remove them. Very counter-intuitive.
* Since 7.5.0 array properties (even optional ones) are initialized with an empty list. This can cause problems if you have a "minItems=1" validation in place. The behavior can be disabled with config option containerDefaultToNull = true, but then also required array properties are initialized to null, which may cause problems in application code that assumes the list to be not null.
* Since 7.7.0 enum values are generated in UPPER_SNAKE_CASE
* Since 7.11.0 a `@JsonCreator` with `@JsonProperty(required="true")` is generated for types with required properties. When the properties are not present during deserialization, a jackson MismatchedInputException is thrown. This requires a [change](https://github.com/belgif/rest-problem-java/issues/167) in the rest-problem-java library to map this to a proper BadRequestProblem.
There's also this related [regression for required properties in a oneOf](https://github.com/OpenAPITools/openapi-generator/issues/20545).

**Reply by pvdbosch** [reply]
We've encountered the change in array behavior as well; seems to be under discussion in https://github.com/OpenAPITools/openapi-generator/issues/18735 . I haven't delved into the config options yet, but I'd expect similar to e.g. string properties:

* JSON absent array property => null in Java, (or input rejected if property is specified as required)
* JSON null array property => null or JsonNullable in Java (or input rejected if property doesn't have "nullable: true")
* JSON empty array => empty array in Java

We're also impacted by the enum values change. It's desired behavior, but backwards-incompatible.

**Comment by pvdbosch** [comment]
For multipart responses, we're finding that generated code doesn't work as-is in Spring bc it doesn't support converting them ( https://github.com/spring-projects/spring-framework/issues/33263 ). We're looking at developing a custom Converter, inspired on [this one](https://github.com/ozonophore/MultipartMessageConverter/blob/main/src/main/java/su/common/util/converter/MultiPartMessageConverter.java).
